### ISSUE TITLE
Establish NPM release process

### ISSUE THREAD

**Issue by danburzo** [mozilla/readability#640]
Now that it seems Readability is getting a steady stream of contributions, I would like to propose a basic process for releasing to npm.

## Steps to release

(These could be included in a _CONTRIBUTING.md_ file)

1. `npm version [patch | minor | major]`, depending on the nature of the changes according to [semver](https://semver.org/). This will bump the version in `package.json` and `package-lock.json` and create a commit and Git tag for the release.
2. `npm publish` to push the release to the npm registry.
3. `git push origin head --follow-tags` to push the new commit and tag to GitHub.
4. optionally, publish a new Release off the newly pushed tag (see _Keeping a changelog_).

## Keeping a changelog

Two widely used approaches here are:

1. A _CHANGELOG.md_ file (GitHub-agnostic)
2. Using the [repository's Releases](https://github.com/mozilla/readability/releases) tab (GitHub-specific)

We can one, or both.

**Comment by gijsk** [comment]
Thanks for filing a ticket for this. Unsure if you're on Mozilla's matrix ; coincidentally I was just asking about this about 10 minutes before you filed this. 😄 

The other part of this is that I still have to manually merge things into the main Firefox repos, for which I have to write bugzilla tickets, attach a patch, get review, make sure tests *there* (for the actual reader mode feature) are still passing, and then land changes. At least tagged releases and tracking a changelog are going to make it easier to track changes there, too.

I have two questions around this:

1. Is it possible to automate more of the steps and/or run them from a web interface as a single process?
2. What's the best practice around *when* you release? After every commit that changes behaviour of the module? Or at regular intervals? Or is deciding that part a bit of a free-for-all?

**Comment by danburzo** [comment]
> Unsure if you're on Mozilla's matrix ; coincidentally I was just asking about this about 10 minutes before you filed this. 😄

I'm not, it was just something that was on my mind for a while, and I got nudged by the recent activity :-)

> 1. Is it possible to automate more of the github/npm steps you outline and/or run them from a web interface as a single process?

It should be possible through a CI/CD pipeline, such as Travis or GitHub Actions. I found [this article](https://cloudfour.com/thinks/how-to-publish-an-updated-version-of-an-npm-package/), which goes into a lot of detail, including options for automation. [And a recent one that popped up in my RSS reader](https://css-tricks.com/how-to-automate-project-versioning-and-releases-with-continuous-deployment/).

> 2. What's the best practice around when you release? After every commit that changes behaviour of the module? Or at regular intervals? Or is deciding that part a bit of a free-for-all?

I don't think there are hard and fast rules about it, no. I would assume at least some level of batching, especially when releasing breaking changes — it's demoralizing to have a dependency jump from `2.0.0` to `7.0.0` since the last time you  checked.

**Comment by dgellow** [comment]
Automation should be possible by adding a task to the file [taskcluster.yml](https://github.com/mozilla/readability/blob/master/.taskcluster.yml) we added earlier today.

I'm not sure what we should add exactly, but we would need at least a new field for the `$match` entry, with something close to:
```yml
    (tasks_for == "github-release"):
      taskId:
        $eval: as_slugid("release_task")
      provisionerId: proj-misc
      workerType: ci
      deadline: {$fromNow: '1 day'}
      payload:
        maxRunTime: 600
        image: node
        command:
          - /bin/bash
          - '--login'
          - '-c'
          - >-
            git clone ${repository} repo && cd repo && git config
            advice.detachedHead false && git checkout ${head_rev} &&
            npm install &&
            npm publish # <== here do whatever is needed to publish the package
      metadata:
        name: Push release
        description: 'Push the release to NPM registry'
        owner: '${event.sender.login}@users.noreply.github.com'
        source: '${event.repository.url}'
```

Of course it depends what parts of the release process should be automated. For example, with Github actions, the draft of a changelog and a release can be automated using https://probot.github.io/apps/release-drafter/. But porting `release-drafter` to TaskCluster would require quite a lot of work.

We can maybe find examples from other projects using TaskCluster, I'm sure others are facing the same problem.

**Comment by dgellow** [comment]
Another approach, the CLI tool [release-it](https://github.com/release-it/release-it) seems to be environment agnostic (so could be run from any CI, or manually) and to handle almost everything I can think of related to the release process.

The interactive version look like this (from their [documentation](https://github.com/release-it/release-it/tree/master/docs/assets)):
![](https://github.com/release-it/release-it/blob/master/docs/assets/release-it-interactive.gif?raw=true)

**Comment by cckang** [comment]
One noob beginner question. I want to learn Readability library by trying to fix some open issues. How should I test my changes? In mozilla code base? I try to load it as an external script source for some loaded page but seems like something is broken. 
If this comment is not appropriate or not suitable to post here. you may delete it. Thanks :)

**Comment by gijsk** [comment]
@cckang yeah, please don't ask in unrelated issues. :-)
You can ping me on matrix for questions - https://chat.mozilla.org/#/room/#fx-desktop-dev:mozilla.org 
To test things, run `npm test` - you can add new tests using the `generate-testcase.js` test script by passing it an alias and a URL to a page you want to use as a testcase.

**Comment by gijsk** [comment]
`release-it` looks pretty good - I'm hoping to make time for this tomorrow, otherwise next week.

**Comment by gijsk** [comment]
> I'm hoping to make time for this tomorrow, otherwise next week.

Haha, haha, ha. 😅

Well, here we are, 2 days before Christmas and 20 days after that comment - but I've finally taken a broom through this stuff.

There's now a `CONTRIBUTING.md` (linked from the readme), and a `CHANGELOG.md` and I copy-pasted some stackoverflow stuff to retro-actively tag 0.3.0 with the correct date, too, and released 0.4.0. So I *think* this is all done, at a rudimentary level.

Obviously there's probably room for improvement in both so please do send PRs if you've spotted typos, things I've missed, or suboptimal choices I've made.

I decided not to do github releases for now, because having a changelog file seemed more portable, and otherwise I didn't think there'd be much point as this isn't consumer-oriented in-and-of-itself (there are tags, for which github also provides convenient archives, and NPM releases, so I'd expect those to work for most people?), but if there is a compelling reason to add them, please file a separate ticket for that.

**Comment by danburzo** [comment]
This is perfect, thank you!
